### ISSUE TITLE
Make floating point numbers look like numbers, not abbreviations

### ISSUE THREAD

**Issue by daniel-v-klein** [PyCQA/pylint#3717]
Python is liberal about what a float looks like, humans have more problems. A leading or trailing . is easy to miss, and with the wrong font or size, can look like a ,, impeding readability (and the text immediately before "impeding" is an example of that).

### Describe the solution you'd like
Require floating point numbers to have leading and trailing digits (that is, there must always be a digit on both sides of the '.')

BAD:  1., 2., .3, 4., 5.1

GOOD: 1.0, 2.0, 0.3, 4.0, 5.1

### Additional context
The Python syntax is unambiguous, but humans are not compilers :-)

**Comment by AWhetter** [comment]
To clarify, the last example in the "BAD" section (`5.1`) is good, correct?

**Comment by daniel-v-klein** [comment]
5.1 is good, .1 and 5. are bad.

**Comment by clavedeluna** [comment]
@Pierre-Sassoulas should this be an extension or base checker?

**Comment by Pierre-Sassoulas** [comment]
Imo, extension with a convention level. To be fair it sounds like creating a pre-commit hook to fix it automatically wouldn't be that much more work.

**Comment by clavedeluna** [comment]
Sounds like a fun project with a fun regex 😄 

So I started looking into this and noticed something critical. `astroid` doesn't distinguish between 1. and 1.0:


```
import astroid

bad = astroid.extract_node("var = 1.")
good = astroid.extract_node("var = 1.0")

good.value.value
> 1.0
bad.value.value
> 1.0
```

I'm not surprised at this behavior at all, and I actually think this is good behavior.... but then how would pylint detect the bad case? Seems to me this is a "won't do" issue and I can write the checker separately and make a pre-commit hook for it for anyone who wants it!

**Comment by daniel-v-klein** [comment]
The floating point values .1 or 1. are both legal values, however the regex
for a float *in the language*, but IMHO should be flagged in a linter.
Example of confusing code:

t = tuple(3.,2.1,.1)      # Adjacent punctuation makes this harder to read
t = tuple(3.0,2.1,0.1)    # Better, no hard-to-read adjacent punctuation...
t = tuple(3.0, 2.1, 0.1)  # Best, with spacing around commas

x = 1.                    # My eyes aren't great any more... W
x = 1,                    # ...which is the tuple and which is the float?

Obviously, fixing this in strings is a problem that cannot/should not be
fixed (IPv4 addresses as the primary example), but in numbers... I think it
improves readability.

**Comment by clavedeluna** [comment]
Indeed that's what I saw, too. @mbyrnepr2 do you think this issue should be closed with this explanation?

**Comment by Pierre-Sassoulas** [comment]
The checker isn't impossible to do, we can also do a TokenChecker. Something like this:

```python
import tokenize

from pylint.checkers import BaseTokenChecker

class FloatChecker(BaseTokenChecker):
    def process_tokens(self, tokens: list[tokenize.TokenInfo]) -> None:
        for idx, (tok_type, string, start, _, line) in enumerate(tokens):
            print(idx, (tok_type, string, start, _, line))
```      
On a.py:
```python
tomato = .5
```
The result of ``pylint a.py`` is the following:
```         
0 (63, 'utf-8', (0, 0), (0, 0), '')
1 (1, 'tomato', (1, 0), (1, 6), 'tomato = .5')
2 (54, '=', (1, 7), (1, 8), 'tomato = .5')
3 (2, '.5', (1, 9), (1, 11), 'tomato = .5')
4 (4, '', (1, 11), (1, 12), '')
5 (0, '', (2, 0), (2, 0), '')
--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)
```

**Comment by DanielNoord** [comment]
If this is a token checker it seems this would have a better fit in `flake8`. Let's focus on the things we're good at :smile:

**Comment by Pierre-Sassoulas** [comment]
There's nothing wrong with token checker in pylint imo. pylint strength is it's versatility : If the AST prevents us from doing a check we have a design that permit to do it anyway. There's no reason to use inference and the node visitor pattern if a check is valuable and correct when using tokens.
